matrix-client/src/Network/Matrix/Internal.hs

```diff
@@ -156,7 +164,67 @@ instance FromJSON MatrixError where
 -- | 'MatrixIO' is a convenient type alias for server response
 type MatrixIO a = MatrixM IO a
 
-type MatrixM m a = m (Either MatrixError a)
+-- | The session record, use 'createSession' to create it.
+data ClientSession = ClientSession
+  { baseUrl :: Text,
+    token :: MatrixToken,
+    manager :: HTTP.Manager
+  }
+
+-- | 'createSession' creates the session record.
+createSession ::
+  -- | The matrix client-server base url, e.g. "https://matrix.org"
+  Text ->
+  -- | The user token
+  MatrixToken ->
+  IO ClientSession
+createSession baseUrl' token' = ClientSession baseUrl' token' <$> mkManager
+
+-- | 'createSession' creates the session record.
+createSessionWithManager ::
+  -- | The matrix client-server base url, e.g. "https://matrix.org"
+  Text ->
+  -- | The user token
+  MatrixToken ->
+  -- | A 'http-client' Manager
+  HTTP.Manager ->
+  ClientSession
+createSessionWithManager = ClientSession
+
+mkRequest :: MonadIO m => Bool -> Text -> MatrixM m HTTP.Request
+mkRequest auth path = do
+  ClientSession {..} <- ask
+  liftIO $ mkRequest' baseUrl token auth path
+
+doRequest :: forall a m. (MonadIO m, FromJSON a) => HTTP.Request -> MatrixM m a
+doRequest request = do
+  ClientSession {..} <- ask
+  MatrixM $ ExceptT $ liftIO $ doRequest' manager request
+
+newtype MatrixM m a = MatrixM { unMatrixM :: ExceptT MatrixError (ReaderT ClientSession m) a }
```

Comment:
IMO we'd better call this `MatrixT` as it's a monad transformer. Also, we can have a interface like:

```haskell
class MonadMatrix m where
  getClientSession :: m ClientSession
  throwMatrixError :: MatrixError -> m a
  performHttpRequest :: FromJSON a => HTTP.Request -> m a
```

 Then `MatrixT` could be a default implementation of `MonadMatrix`:

```haskell
instance MonadIO m => MonadMatrix (MatrixT m) where
  getClientSession = ask
  throwMatrixError = throwError
  performHttpRequest = ...
```

and users may create their own monads with or wihout `MatrixT`.

Reply:
Presently, this is going to be tricky when using both a regular session and a identity session, I think we need to implement a ClientSession that support both endpoints, otherwise I think it's going to be tricky to implement such Monad for both use-case.